File: docs/palindromes-in-peg/fpp_cost.py
"""Structural cost bounds for the finite FPP tables, in local instructions.

The control-graph factor is computed. The resource bounds are the symbolic
tape-motion ledger documented in FPP_COST.md, not fitted trace maxima. The
observer below makes each ledger obligation separately testable; it supplies
no runtime operation to the finite machine.
"""
from dataclasses import dataclass

from fpp_finite import A, B, C, S, T, BACK, FRONT, build_program


def successors(row):
  if row[0] == "read":
    return tuple(row[2].values())
  return () if row[0] == "halt" else (row[-1],)


def is_cut(row):
  return row[0] in ("move", "emit", "halt") or row[:2] == ("read", FRONT)


def cut_distance(program):
  """Maximum instructions through the next charged instruction, inclusive.

  Cutting moves, queue-front reads, outputs and halt makes the remaining
  graph acyclic. A non-cut cycle is an error, never a guessed finite bound.
  """
  visiting, distances = set(), {}

  def distance(q):
    if q in distances:
      return distances[q]
    if q in visiting:
      raise ValueError("uncharged cycle in finite FPP control graph")
    visiting.add(q)
    row = program.code[q]
    value = 1 if is_cut(row) else 1 + max(distance(k) for k in successors(row))
    visiting.remove(q)
    distances[q] = value
    return value

  return max(distance(q) for q in range(len(program.code)))


@dataclass(frozen=True)
class Affine:
  slope: int
  intercept: int = 0

  def __call__(self, size):
    if type(size) is not int or size < 0:
      raise ValueError("nonnegative integral size required")
    return self.slope * size + self.intercept


MOVE_BOUNDS = tuple(Affine(a, b) for a, b in (
  (4, 0), (1, 0), (6, 1), (7, 1), (2, 0), (4, 0), (4, 0)))
FRONT_READS = Affine(4)
EMITS = Affine(1)
PREPARATION = Affine(24, 42)
DP_SCAN = Affine(18, 25)


def bounds(alphabet="abs"):
  factor = cut_distance(build_program(alphabet + "#"))
  moves = Affine(sum(b.slope for b in MOVE_BOUNDS),
                 sum(b.intercept for b in MOVE_BOUNDS))
  cuts = Affine(moves.slope + FRONT_READS.slope + EMITS.slope,
                moves.intercept + 1)  # the final halt
  kernel = Affine(factor * cuts.slope, factor * cuts.intercept)
  # MARKS follows every A movement. Its emit-to-write replacement costs
  # exactly one instruction, just as the original emit does.
  marked_kernel = Affine(kernel.slope + MOVE_BOUNDS[A].slope, kernel.intercept)
  marked = Affine(2 * marked_kernel.slope + PREPARATION.slope,
                  marked_kernel.slope + marked_kernel.intercept + PREPARATION.intercept)
  # SECOND duplicates MARKS: setup is 3m+4, then A moves plus emits on
  # N=2m+1. The FPP halt is replaced by one read, with the same unit cost.
  duplicate = Affine(3 + 2 * (MOVE_BOUNDS[A].slope + EMITS.slope),
                     4 + MOVE_BOUNDS[A].slope + EMITS.slope)
  dp = Affine(marked.slope + duplicate.slope + DP_SCAN.slope,
              marked.intercept + duplicate.intercept + DP_SCAN.intercept)
  return dict(factor=factor, kernel=kernel, marked=marked, dp=dp)


class CostObservation:
  def __init__(self):
    self.moves = [0] * 7
    self.front_reads = self.emits = self.halts = self.steps = self.cuts = 0

  def observe(self, row):
    self.steps += 1
    self.cuts += int(is_cut(row))
    if row[0] == "move":
      self.moves[row[1]] += 1
    self.front_reads += int(row[:2] == ("read", FRONT))
    self.emits += int(row[0] == "emit")
    self.halts += int(row[0] == "halt")

  def check(self, size, factor):
    for tape, bound in enumerate(MOVE_BOUNDS):
      assert self.moves[tape] <= bound(size), (tape, size, self.moves, bound)
    assert self.front_reads <= FRONT_READS(size), (size, self.front_reads)
    assert self.emits <= EMITS(size) and self.halts == 1
    assert self.steps <= factor * self.cuts, (self.steps, factor, self.cuts)

File: scala/pal/src/main/scala/pal/FppCost.scala
package pal

import scala.collection.mutable

import FppFinite.*
import FppFinite.Instruction.*

/** Structural cost bounds for the finite FPP tables, in local instructions.
  *
  * The control-graph factor is computed. The resource bounds are the symbolic
  * tape-motion ledger documented in FPP_COST.md, not fitted trace maxima. The
  * observer below makes each ledger obligation separately testable; it supplies
  * no runtime operation to the finite machine.
  *
  * Python 原典: `fpp_cost.py`。
  */
object FppCost {

  def successors(row: Instruction): Seq[Int] = row match {
    case Read(_, choices) => choices.values.toSeq
    case Halt => Nil
    case Emit(next) => Seq(next)
    case Move(_, _, next) => Seq(next)
    case Write(_, _, next) => Seq(next)
  }

  /** Instructions that are charged to a tape-motion resource: moves, queue-front reads, outputs, halt. */
  def isCut(row: Instruction): Boolean = row match {
    case Move(_, _, _) | Emit(_) | Halt => true
    case Read(FRONT, _) => true
    case _ => false
  }

  /** Maximum instructions through the next charged instruction, inclusive.
    *
    * Cutting moves, queue-front reads, outputs and halt makes the remaining
    * graph acyclic. A non-cut cycle is an error, never a guessed finite bound.
    */
  def cutDistance(program: Program): Int = {
    val visiting = mutable.HashSet.empty[Int]
    val distances = mutable.HashMap.empty[Int, Int]

    def distance(q: Int): Int = {
      distances.get(q) match {
        case Some(d) => d
        case None =>
          if (visiting.contains(q)) {
            throw new IllegalArgumentException("uncharged cycle in finite FPP control graph")
          }
          visiting += q
          val row = program.instruction(q)
          val value = if (isCut(row)) { 1 } else { 1 + successors(row).map(distance).max }
          visiting -= q
          distances(q) = value
          value
      }
    }

    program.code.indices.map(distance).max
  }

  /** `slope * size + intercept`, defined on nonnegative sizes only. */
  final case class Affine(slope: Int, intercept: Int = 0) {
    def apply(size: Int): Int = {
      if (size < 0) {
        throw new IllegalArgumentException("nonnegative integral size required")
      }
      slope * size + intercept
    }
  }

  /** Per-tape move bounds, indexed by A, B, C, S, T, BACK, FRONT. */
  val MOVE_BOUNDS: Vector[Affine] = Vector(
    Affine(4, 0), Affine(1, 0), Affine(6, 1), Affine(7, 1), Affine(2, 0), Affine(4, 0), Affine(4, 0))
  val FRONT_READS: Affine = Affine(4)
  val EMITS: Affine = Affine(1)
  val PREPARATION: Affine = Affine(24, 42)
  val DP_SCAN: Affine = Affine(18, 25)

  /** Python の `bounds()` が返す dict。 */
  final case class Bounds(factor: Int, kernel: Affine, marked: Affine, dp: Affine)

  def bounds(alphabet: String = "abs"): Bounds = {
    val factor = cutDistance(buildProgram(alphabet + "#"))
    val moves = Affine(MOVE_BOUNDS.map(_.slope).sum, MOVE_BOUNDS.map(_.intercept).sum)
    val cuts = Affine(moves.slope + FRONT_READS.slope + EMITS.slope,
      moves.intercept + 1) // the final halt
    val kernel = Affine(factor * cuts.slope, factor * cuts.intercept)
    // MARKS follows every A movement. Its emit-to-write replacement costs
    // exactly one instruction, just as the original emit does.
    val markedKernel = Affine(kernel.slope + MOVE_BOUNDS(A).slope, kernel.intercept)
    val marked = Affine(2 * markedKernel.slope + PREPARATION.slope,
      markedKernel.slope + markedKernel.intercept + PREPARATION.intercept)
    // SECOND duplicates MARKS: setup is 3m+4, then A moves plus emits on
    // N=2m+1. The FPP halt is replaced by one read, with the same unit cost.
    val duplicate = Affine(3 + 2 * (MOVE_BOUNDS(A).slope + EMITS.slope),
      4 + MOVE_BOUNDS(A).slope + EMITS.slope)
    val dp = Affine(marked.slope + duplicate.slope + DP_SCAN.slope,
      marked.intercept + duplicate.intercept + DP_SCAN.intercept)
    Bounds(factor, kernel, marked, dp)
  }

  /** Counts the charged instructions of one execution and checks them against the ledger. */
  final class CostObservation {
    val moves: Array[Int] = Array.fill(7)(0)
    var frontReads: Int = 0
    var emits: Int = 0
    var halts: Int = 0
    var steps: Int = 0
    var cuts: Int = 0

    def observe(row: Instruction): Unit = {
      steps += 1
      if (isCut(row)) { cuts += 1 }
      row match {
        case Move(tape, _, _) => moves(tape) += 1
        case Read(FRONT, _) => frontReads += 1
        case Emit(_) => emits += 1
        case Halt => halts += 1
        case _ => ()
      }
    }

    def check(size: Int, factor: Int): Unit = {
      for ((bound, tape) <- MOVE_BOUNDS.zipWithIndex) {
        assert(moves(tape) <= bound(size), (tape, size, moves.toVector, bound))
      }
      assert(frontReads <= FRONT_READS(size), (size, frontReads))
      assert(emits <= EMITS(size) && halts == 1)
      assert(steps <= factor * cuts, (steps, factor, cuts))
    }
  }
}
